core/include/vecmem/memory/conditional_memory_resource.hpp

```diff
@@ -0,0 +1,48 @@
+/**
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include <cstddef>
+#include <functional>
+
+#include "vecmem/memory/memory_resource.hpp"
+
+namespace vecmem {
+/**
+ * @brief This memory resource conditionally allocates memory. It is
+ * constructed with a predicate function that determines whether an allocation
+ * should succeed or not.
+ *
+ * This resource can be used to construct complex conditional allocation
+ * schemes.
+ */
+class conditional_memory_resource : public memory_resource {
+public:
+    /**
+     * @brief Constructs the conditional memory resource.
+     *
+     * @param[in] upstream The upstream memory resource to use.
+     * @param[in] pred The predicate function that determines whether the
+     * allocation should succeed.
+     */
+    conditional_memory_resource(
+        memory_resource& upstream,
+        std::function<bool(std::size_t, std::size_t)> pred);
+
+private:
+    virtual void* do_allocate(std::size_t, std::size_t) override;
+
+    virtual void do_deallocate(void* p, std::size_t, std::size_t) override;
+
+    virtual bool do_is_equal(const memory_resource&) const noexcept override;
+
+    memory_resource& m_upstream;
```

Comment:
I'm going to have to disagree with you on this one. The idea behind `reference_wrapper` is to provide a reseatable reference type. The main reason it exists is to allow you to put references in standard containers, which is otherwise impossible. In this case, you don't ever want to reseat the reference in a memory resource, that would be extremely dangerous. In this case, that would mean that you can allocate memory with one resource, reseat the upstream reference, and then deallocate that same memory with a different resource. This could lead to huge problems.

Reply:
:thinking: So you **do** want to prevent these types of objects from being copied. Why?

If somebody allocates memory using one instance of `vecmem::conditional_memory_resource`, but then that instance is destructed (while being copied somewhere else), that's just a coding error on the user's side. We can't really prevent the users from destructing memory resource objects too early anyway.

Every time that I dealt with variadic templates (which we may end up doing on top of these memory resources at one point), I always had some number of issues coming from non-copyable types. As those are generally just much more tricky to handle in "generic code". This is mainly why I like to keep classes that can in principle be copied, to be allowed to be copied.

Reply:
Fine. I agree that we don't want to update existing objects with assignment...